draft-ietf-quic-qpack.md

```diff
@@ -95,17 +95,29 @@ reintroduced by HPACK when the loss includes a HEADERS frame.
 
 ## Avoiding Head-of-Line Blocking in HTTP/QUIC {#overview-hol-avoidance}
 
-In the example above, the second stream contained a reference to data
-which might not yet have been processed by the recipient.  Such references
-are called "vulnerable," because the loss of a different packet can keep
-the reference from being usable.
-
-The encoder can choose on a per-header-block basis whether to favor higher
+Because QUIC does not guarantee order between data on different streams, a
+reference might be received on a request stream to an entry which is not yet in
+the dynamic table. References are called "vulnerable" when they refer to entries
+whose receipt has not been confirmed by the peer, because loss or reordering can
+keep the reference from being immediately usable.  If the reference cannot be
+immediately processed on receipt, the stream is considered "blocked."
```

Comment:
I *still* don't like "vulnerable" and I think that I understand better now.  The problem is that you are putting the cart before the horse.  The problem is that the decoder might block.  The focus on the encoder side is wrong.

Hence, I suggest, without the word "vulnerable" at all:

> [The QUIC is not ordered preface]. A decoder that receives a reference to a dynamic table entry that has not been received is "blocked".
>
> An encoder can decide whether to risk having a header block becoming blocked at a decoder.  Compression efficiency can be improved by referencing dynamic table entries that are still in transit, but if there is packet loss or reordering a decoder can become blocked.  An encoder avoids the risk of blocking by only referencing dynamic table entries that have been acknowledged, though this means using literals, which make the header block larger.  Larger header blocks can result in an encoder become blocked on congestion or flow control limits.
>
> A decoder can permit the possibility of blocking references using the SETTINGS_QPACK_BLOCKING_REFERENCES setting.  This setting sets an upper bound on the number of header blocks that can reference unacknowledged entries in the dynamic table.  An encoder respects this setting when deciding whether to reference unacknowledged dynamic table entries.
>
> Every header block contains a Largest Reference (see ...).  This is used to determine the minimum number of entries need to be present in the dynamic table before the header block can be successfully decoded.  A decoder becomes blocked if the Largest Reference is greater than the number of dynamic table entries that it has received.
>
> An encoder MUST limit the number of unacknowledged header blocks that reference unacknowledged dynamic table entries to the value of the SETTINGS_...  A decoder that is blocked on a number of header blocks equal to the limit it advertised SHOULD treat receipt of an additional blocked header block [awkward] as a stream error of type HTTP_...

Or something like that.

Reply:
I think this is a workable direction.  We should be consistent about blocked vs blocking (I prefer blocked), and what is blocked - is it the stream, header block, or reference?

A single header block may contain many references to unacknowledged entries, so it's not blocked references the decoder is counting, but header blocks.  It seems logical that the entire stream should be blocked until the blocked header block is decoded, but perhaps the text needs to be more explicit about it.

All of this to say that I like Mike's suggestion: SETTINGS_QPACK_BLOCKED_STREAMS.

Reply:
Yes, I think that the unit is header blocks.  A decoder is blocked on the decoding of a header block.  

We can generalize that to streams, since the correspondence is perfect to streams: a blocked header block is a blocked stream.  In fact, we should...

Thinking more about that, if an endpoint is silly and they read a blocked header block, then concentrating on header blocks might be wrong.  A message with trailers has two header blocks and if both are blocked, you might expect the encoder to only account for one lot of blocking.  But a silly decoder might find itself blocked on both header and trailer at the same time.  Specifying this as the number of streams that have a decoder blocked on the decoding of a header block with unresolved references to the dynamic table is ideal.

Reply:
I agree that the unit should be streams; the decoder SHOULD NOT be reading even the body of the header block, which means it won't see any further header blocks on that stream.  A cautious encoder might choose to count blocks for simplicity, but the decoder's count will almost certainly be one per stream.  A silly decoder might read further in an attempt to over-optimize, and specifying the blocking count to be streams helps keep them from blowing up.